rules-xml/impl/src/main/java/org/jboss/windup/rules/apps/xml/xml/ValidateXmlFilesRuleProvider.java

```diff
@@ -0,0 +1,170 @@
+import javax.xml.validation.Schema;
+import javax.xml.validation.SchemaFactory;
+import javax.xml.validation.Validator;
+import java.io.FileNotFoundException;
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.net.UnknownHostException;
+import java.util.Collections;
+
+/**
+ * A rule provider validating all of the xml files and registering the classification in case the xml file is not valid.
+ *
+ * @author <a href="mailto:mbriskar@gmail.com">Matej Briskar</a>
+ */
+public class ValidateXmlFilesRuleProvider extends AbstractRuleProvider
+{
+
+    public static final String XMLSCHEMA_URL = "http://www.w3.org/2001/XMLSchema-instance";
+    public static final String XMLSCHEMA_ATTRIBUTE_NAME = "schemaLocation";
+    public static final String NOT_VALID_XML_TAG = "Not valid XML";
+
+    SchemaFactory schemaFactory = SchemaFactory
+                .newInstance(XMLConstants.W3C_XML_SCHEMA_NS_URI);
+
+    public ValidateXmlFilesRuleProvider()
+    {
+        this(SchemaFactory.newInstance(XMLConstants.W3C_XML_SCHEMA_NS_URI));
+    }
+
+    public ValidateXmlFilesRuleProvider(SchemaFactory schemaFactory)
+    {
+        super(MetadataBuilder.forProvider(ValidateXmlFilesRuleProvider.class)
+                    .setPhase(MigrationRulesPhase.class)
+                    .setHaltOnException(true));
+        this.schemaFactory=schemaFactory;
+    }
+
+    @Override
+    public Configuration getConfiguration(GraphContext context)
+    {
+        return ConfigurationBuilder.begin()
+                    .addRule()
+                    .when(new ProcessingIsOnlineGraphCondition().and(Query.fromType(XmlFileModel.class)))
+                    .perform(new ValidateAndRegisterClassification());
+    }
+
+    private class ValidateAndRegisterClassification extends AbstractIterationOperation<XmlFileModel>
+    {
+        @Override
+        public void perform(GraphRewrite event, EvaluationContext context, XmlFileModel sourceFile)
+        {
+            XmlFileService xmlFileService = new XmlFileService(event.getGraphContext());
+            Document document = xmlFileService.loadDocumentQuiet(context, sourceFile);
+            String xsdUrl = getXSDSchemaLocationUrl(document);
+            try
+            {
+                if(xsdUrl != null && !xsdUrl.isEmpty() && !isXmlValid(sourceFile,xsdUrl)) {
+                    ClassificationService classificationService = new ClassificationService(event.getGraphContext());
+                    ClassificationModel model =classificationService.attachClassification(context, sourceFile, XmlFileModel.NOT_VALID_XML,
+                                null);
+                    TagSetService tagSetService = new TagSetService(event.getGraphContext());
+                    model.setTagModel(tagSetService.getOrCreate(event, Collections.singleton(NOT_VALID_XML_TAG)));
+                }
+            }
+            catch (MalformedURLException e)
+            {
+                e.printStackTrace();
```

Comment:
maybe, however I was not able to reproduce this exception so I am not sure when it happens.. I may look into it more